### PR TITLE
feat(HACBS-1233): add the binding controller

### PR DESCRIPTION
**Pull request by dirgim**
* Add binding controller to act on integration test deployments on ephemeral envs
* Add functions that can detect a SEB with a finished deployment
* Ensure creation of integration test pipelines which utilize ephemeral envs
* Add unit tests for the binding controller
* Add unit tests for updated code in other packages

### CODE REVIEW

**Review by hongweiliu17 — COMMENTED** [review]
I learn and review the code, I feel it looks good to me other than some minor comments.

File: controllers/binding/binding_adapter.go [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+	applicationapiv1alpha1 "github.com/redhat-appstudio/application-api/api/v1alpha1"
+	"github.com/redhat-appstudio/integration-service/api/v1alpha1"
+	"github.com/redhat-appstudio/integration-service/controllers/results"
+	"github.com/redhat-appstudio/integration-service/gitops"
+	"github.com/redhat-appstudio/integration-service/helpers"
+	"github.com/redhat-appstudio/integration-service/tekton"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// Adapter holds the objects needed to reconcile a Release.
```
[the remark this answers is not in the document]
Comment (dirgim) [reply controllers/binding/binding_adapter.go:33]: Good catch, thanks!

File: controllers/binding/binding_adapter.go [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+				a.logger.Error(err, "Failed to create pipelineRun for snapshot, environment and scenario")
+				return results.RequeueOnErrorOrStop(err)
+			}
+		}
+	}
+
+	return results.ContinueProcessing()
+}
+
+// createIntegrationPipelineRunWithEnvironment creates and returns a new integration PipelineRun. The Pipeline information and the parameters to it
```
[the remark this answers is not in the document]
Comment (dirgim) [reply controllers/binding/binding_adapter.go:95]: Good point, will adjust.

File: controllers/binding/binding_controller.go [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+		if errors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+
+		return ctrl.Result{}, err
+	}
+
+	application, err := r.getApplicationFromSnapshotEnvironmentBinding(ctx, snapshotEnvironmentBinding)
+	if err != nil {
+		logger.Error(err, "Failed to get Application for ",
```
[the remark this answers is not in the document]
Comment (dirgim) [reply controllers/binding/binding_controller.go:76]: Another good catch, appreciate it!

**Review by jsztuka — APPROVED** [review]
lgtm

File: controllers/binding/binding_controller.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (dirgim) [reply controllers/binding/binding_controller.go:76]: Updated the log message

File: controllers/binding/binding_adapter.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (dirgim) [reply controllers/binding/binding_adapter.go]: Updated the doc string, please check if it looks good now

**Review by hongweiliu17 — APPROVED** [review]
LGTM

**Pull request merged**